src/modules/battery_resistance_est/battery_resistance_est.cpp

```diff
@@ -0,0 +1,287 @@
+
+	return true;
+}
+
+Vector<float, 4> InternalRes::extract_ecm_parameters()
+{
+
+	Vector<float, 4> ecm_params;
+
+	ecm_params(0) = _param_est(0); //_r_steady_state
+	ecm_params(1) = (_param_est(1) / _param_est(2)) - _param_est(0); //_r_transient
+	ecm_params(2) = _param_est(3) / _param_est(2); //_voltage_open_circuit
+
+	//calculate bat1_r_internal
+	ecm_params(3) = (_voltage_filtered_v - ecm_params(2)) / (-_current_filtered_a); //_internal_resistance_est
+
+	//logging
+	_inter_res.r_s = ecm_params(0);
+	_inter_res.r_t = ecm_params(1);
+	_inter_res.v_oc = ecm_params(2);
+	_inter_res.r_internal_est = ecm_params(3);
+
+	return ecm_params;
+}
+
+
+void InternalRes::update_internal_resistance(const float voltage_estimation_error,
+		const Vector<float, 4> &esm_params_est)
+{
+
+	//store best esimate
+	if ((abs(voltage_estimation_error)) <= abs(_best_prediction_error)) {
+		_best_ecm_params_est = esm_params_est;
+		_best_prediction_error = voltage_estimation_error;
+
+	} else if (_best_prediction_error_reset) {
+		_best_ecm_params_est = esm_params_est;
+		_best_prediction_error = voltage_estimation_error;
+		_best_prediction_error_reset = false;
+	}
+
+	//clamp BAT1_R_INTERNAL
+	if (_best_ecm_params_est(3) > _param_inter_res_est_max.get()) {
+		_best_ecm_params_est(3) = _param_inter_res_est_max.get();
+
+	} else if (_best_ecm_params_est(3) < _param_inter_res_est_min.get()) {
+		_best_ecm_params_est(3) = _param_inter_res_est_min.get();
+	}
+
+	const float time_since_param_update = (hrt_absolute_time() - _last_param_update_time) / 1e6f;
+
+	// publish new bat1_r_internal only periodically
+	if (time_since_param_update >= _param_inter_res_update_period.get()) {
+
+		//BAT${i}_R_INTERNAL  increment: 0.01
+		_best_ecm_params_est(3) = round(_best_ecm_params_est(3) * 100) / 100;
+
+		//Set px4 Internal resistance using simplified Rint Model
+		_inter_res.best_r_internal_est = _best_ecm_params_est(3);
+		_last_param_update_time = hrt_absolute_time();
+		_best_prediction_error_reset = true;
+	}
+}
+
+float InternalRes::predict_voltage(const float dt)
+{
+
+	//process _signal
+	_signal(0) = -(_battery_status.current_filtered_a - _current_filtered_a_prev)
+		     / ((_battery_status.timestamp - _battery_time_prev) / 1e6f); //central difference method
+	_signal(1) = -_current_filtered_a;
+	_signal(2) = -_voltage_estimation;
+	_signal(3) = 1.f;
+
+	// Predict the voltage using the learned adaptive model
+	_voltage_estimation += (_param_est.transpose() * _signal * dt)(0, 0);
+
+	const float voltage_estimation_error = _voltage_filtered_v - _voltage_estimation;
+
+	_voltage_estimation += _lambda * dt * voltage_estimation_error;
+
+	//logging
+	_inter_res.voltage_estimation = _voltage_estimation;
+	_inter_res.voltage_estimation_error = abs(voltage_estimation_error);
+
+	_inter_res.param_est[0] = _param_est(0);
+	_inter_res.param_est[1] = _param_est(1);
+	_inter_res.param_est[2] = _param_est(2);
+	_inter_res.param_est[3] = _param_est(3);
```

Comment:
```suggestion
	_param_est.copyTo(_inter_res.param_est);
```

Reply:
Thanks again for the suggestions!

Tested squashed changes on top of V1.11.0
Here are 3 replays. (Initial behavior depends on initial params.)
https://review.px4.io/plot_app?log=7d5a651e-27d8-487a-98cc-204d4811e591
https://review.px4.io/plot_app?log=e92d7bf2-02ff-4314-8613-8a0ab127e2c5
https://logs.px4.io/plot_app?log=a4eb35b2-7b69-408b-94b8-4eabff399ad6

Not yet, I will test on a pixhawk 4.

Reply:
@bresch I tested it on a drone. I made a few changes to params so they are based on per cell values and initialized the time variables which I think were causing the nan estimates on the hardware.

here is a log that looks okay to me... voltage open circuit corrects and is always more than v_load. (props were off, tested on x500)

on 5fb11c1e3ed7493590ae694fb5f2cee65a049de8
https://review.px4.io/plot_app?log=07fd1508-9cdb-4d49-a0fa-177f34071cdf

![image](https://user-images.githubusercontent.com/50725049/107156427-8b5f3380-697e-11eb-9bff-97c548b4805f.png)
![image](https://user-images.githubusercontent.com/50725049/107156443-a5991180-697e-11eb-8407-dd4ccfe3cdce.png)